File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * n
        right = [1] * n

        for i in range(n - 1):
            left[i + 1] = left[i] * nums[i]
        
        for i in range(n - 1, 0, -1):
            right[i - 1] = right[i] * nums[i]
        
        res = []

        for i in range(n):
            res.append(left[i] * right[i])
        
        return res

File: Data Structures & Algorithms/sliding-window-maximum/submission-2.cpp
class Solution 
{
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) 
    {
        // monotonic decreasing (storing index instead of value)
        deque<int> dq;

        int L = 0;
        int n = nums.size();
        vector<int> res;
        for (int R = 0; R < n; ++R)
        {
            // maintain monotonic decrasing deque
            while (!dq.empty() && nums[dq.back()] < nums[R])
            {
                dq.pop_back();
            }
            dq.push_back(R);

            if (L > dq.front())
            {
                dq.pop_front();
            }

            if (R + 1 >= k)
            {
                res.push_back(nums[dq.front()]);
                L++;
            }
        }

        return res;
    }
};
